plugins/uefi-dbx/fu-uefi-dbx-plugin.c

```diff
@@ -48,19 +72,67 @@ fu_uefi_dbx_plugin_init(FuUefiDbxPlugin *self)
 {
 }
 
+static void
+fu_uefi_dbx_plugin_finalize(GObject *object)
+{
+	FuUefiDbxPlugin *self = FU_UEFI_DBX_PLUGIN(object);
+	if (self->snapd_notifier != NULL) {
+		g_object_unref(self->snapd_notifier);
+		self->snapd_notifier = NULL;
+	}
+
+	G_OBJECT_CLASS(fu_uefi_dbx_plugin_parent_class)->finalize(object);
+}
+
+static gboolean
+fu_uefi_dbx_plugin_snapd_notify_init(FuUefiDbxPlugin *self, GError **error)
+{
+	g_autoptr(FuUefiDbxSnapdNotifier) obs = fu_uefi_dbx_snapd_notifier_new();
+
+	if (!fu_uefi_dbx_snapd_notifier_dbx_manager_startup(obs, error))
+		return FALSE;
+
+	g_set_object(&self->snapd_notifier, obs);
+	return TRUE;
+}
+
 static void
 fu_uefi_dbx_plugin_constructed(GObject *obj)
 {
 	FuPlugin *plugin = FU_PLUGIN(obj);
 	fu_plugin_add_rule(plugin, FU_PLUGIN_RULE_METADATA_SOURCE, "uefi_capsule");
 	fu_plugin_add_firmware_gtype(plugin, NULL, FU_TYPE_EFI_SIGNATURE_LIST);
 	fu_plugin_add_device_gtype(plugin, FU_TYPE_UEFI_DBX_DEVICE);
+
+	if (fu_snap_is_in_snap()) {
+		FuUefiDbxPlugin *self = FU_UEFI_DBX_PLUGIN(plugin);
+		g_autoptr(GError) error_local = NULL;
+		/* only enable snapd integration if running inside a snap */
+		if (!fu_uefi_dbx_plugin_snapd_notify_init(FU_UEFI_DBX_PLUGIN(obj), &error_local)) {
+			/* specific error code if relevant APIs are not present and thus
+			 * integration cannot be supported */
+			self->snapd_integration_supported =
+			    !g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED);
+
+			g_info("snapd integration non-functional: %s", error_local->message);
+		} else {
+			g_info("snapd integration enabled ");
+			self->snapd_integration_supported = TRUE;
+		}
+	} else {
+		/* TODO figure out non-snap scenarios */
+		g_info("snapd integration outside of snap is not supported");
+	}
```

Comment:
really you're talking about 3 cases AFAICT.
1. fwupd in a snap
2. fwupd outside of a snap on a system without snapd
3. fwupd outside of a snap on a system with snapd.

Shouldn't that message only show up on case 3?  Would it be harmful to send a "ping" type of message with `fu_uefi_dbx_snapd_notifier_simple_req` in this case to an endpoint?  Maybe a get version command?  I'd expect with a short timeout it wouldn't matter for most systems.

Reply:
I think this is mostly correct, but the problem is that cases 2 and 3 may be indistinguishable if the snapd socket isn't available at a given instant when a request is made. In theory snapd ships with a systemd socket unit, which should lead to the service activation when accessed. However, there may be scenarios when there is a significant delay before response is received, or the socket is briefly gone (eg. snapd package is being updated, package manager briefly stops the socket unit).

To keep things simple, case 1 is what I'm focusing on right now, since it's relatively simple and we already know as a fact that if fwupd is inside a snap, snapd must be around as well.

Ideally, as I mentioned in the other PR, the main discriminator here would be presence of Ubuntu-like-FDE, which is detected by inspecting volumes and partitions, making it completely independent of snapd socket presence. Much like with bitlocker and the way fwupd shows additional warning that updating DBX may result in needing to use recovery keys. With proper detection in place, the snapd integration gets simplified to 2 cases, with or without Ubuntu-FDE. The goal is to make integration mandatory iff FDE is present.

In essence, I would like for this to not have any impact on folks who aren't part of the use case, be sufficiently robust for folks who are and finally for myself to not propose a PR with 3kloc added diff :smile: which would obviously make reviewing much more time consuming. Hope this makes sense.

Reply:
Yeah that makes sense.  I think what you're doing for now is fine.